### ISSUE TITLE
With DD 2.2.0.0 windows backed k8s volume no longer working to hold state for RabbitMQ

### ISSUE THREAD

**Issue by allanbarklie** [docker/for-win#5665]
- [X ] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior

It is possible to map a k8s local persistent volume->persistent volume claim->volume mount backed by a windows folder to /var/lib/rabbitmq/mnesia/ inside a rabbit mq container - to store session state.

### Actual behavior
On 2.2.0.0 RabbitMQ exits during startup with a file access error.

### Information

  - Windows Version: Microsoft Windows 10 Enterprise Version 10.0.18363 Build 18363
  - Docker Desktop Version: 2.2.0.0
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM:
Default Docker Desktop install using kubernetes

Works fine up to 2.1.05,.

The error occurs during an attempt to write a config file to a subfolder of the volume.
My understanding is that some erlang code outputs a temp file in that folder and then renames the file.
The temp file is written but the rename fails with ebusy.
erlang docs suggest that erlang will usually not rename a file if it is open for access elsewhere

The error came from the call to rabbit_file:write_term_file in the function write_cluster_status,
here https://github.com/rabbitmq/rabbitmq-server/blob/master/src/rabbit_node_monitor.erl.

This has worked for as long time across many versions of docker desktop.

### Steps to reproduce the behavior

  1. Pull a rabbitmq image e.g. FROM rabbitmq:3.8.2-management
  2. Setup some local k8s storage, example
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: local-cache-storage
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: message-queue-cache-volume
spec:
  capacity:
    storage: 1000Mi
  accessModes:
  - ReadWriteOnce
  persistentVolumeReclaimPolicy: Retain
  storageClassName: local-cache-storage
  local:
    path: /C/volumes/ServerStorage/Messages
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/hostname
          operator: In
          values:
          - docker-desktop
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: message-queue-pv-claim-1
  namespace: my-namespace
spec:
  accessModes:
  - ReadWriteOnce
  storageClassName: local-cache-storage
  resources:
    requests:
      storage: 1000Mi
  volumeName:
    message-queue-cache-volume
---
3. create a deployment with a rabbitmq pod including 
volumeMounts:
        - name: message-queue-volume
          mountPath: /var/lib/rabbitmq/mnesia/
4. Observe the container fail with a file access error

**Comment by illinar** [comment]
We have observed similar problems with other products (e.g. databases) that perform non-trivial amount of file system calls, including file and folder renames.

You don't experience problems at very small loads, but it can be routinely reproduced once number of operations increases. Typically manifests as access denied with this `another process on the host has locked one of the files` or similar message in the Docker log.

We have performed various experiments on several different setups (including newly installed Windows). Confirmed that it is not AV related and according to _procmon_ there is no "other process" accessing the files other than _com.docker.backend_

Downgrading to 2.1.0.5 immediately resolves the problem. We came to a conclusion that it must be related to new gRPC FUSE implementation.

**Comment by KristofKlein** [comment]
We have the same observation! with 2.2.0.0 we are no longer able to start our RabbitMQ instances with a persistent volume. Neither Kubernetes nor straight docker run.

**Comment by mat007** [comment]
We have a new build that fixes several errors related to file sharing. You can get it from https://download-stage.docker.com/win/stable/42716/Docker%20Desktop%20Installer.exe.
I’d be nice if you could give it a go and report whether this works for you or not. If not could you upload a diagnostics?

**Comment by KristofKlein** [comment]
I don't know how to produce you the diagnostics but I can tell you - it still does not work. To simplify your testing it does not even need K8s to reproduce the issue:
`docker run -it -v c:\tmp\mq:/var/lib/rabbitmq/mnesia rabbitmq:management`

already fails as well with {"init terminating in do_boot",{error,{could_not_write_file,"/var/lib/rabbitmq/mnesia/rabbit@0f2b33ace7ee/cluster_nodes.config",ebusy}}}

**Comment by mat007** [comment]
I just ran this and it seems to be working for me (I’m running the version from the link I posted a couple of messages above), it ends with
```
…
2020-02-11 09:19:02.296 [info] <0.8.0> Server startup complete; 3 plugins started.
 * rabbitmq_management
 * rabbitmq_management_agent
 * rabbitmq_web_dispatch
```

To create a diagnostics open the Troubleshoot window from the systray menu and click Run Diagnostics, then upload it and paste the ID here.

**Comment by KristofKlein** [comment]
there you go: 33A5D7E3-C3A5-41BE-A075-0B8B20C275B8/20200211093804 

what Windows Version are you on? I'm on 1903 (18362.592)

**Comment by mat007** [comment]
19955.1001 but I don’t expect this should make a difference.

**Comment by mat007** [comment]
Ah sorry I hadn’t realized I had enabled WSL 2 …
I can confirm I reproduced the issue with the LinuxVM based file sharing, thanks!

**Comment by allanbarklie** [comment]
Hi,

Thanks for working this through.  I am just catching up - and was about to
ask the wsl2 question :)

Cheers,
Allan.

**Comment by allanbarklie** [comment]
Hi,

Also I can confirm that I still reproduce the issue with 2.2.0.3 (the new
version you linked to).
My windows version is: Version 10.0.18363 Build 18363

Thanks,
Allan.

**Comment by cwenzelg** [comment]
We experience the same issue. After upgrading to 2.2.0.3 we get a lot of timeout and can't read files error in our systems. They appear on different files on random and seem to connect with a higher request rate to mounted files on windows. Downgrading also solves the issue.

**Comment by KristofKlein** [comment]
any news here?!

**Comment by stephen-turner** [comment]
This is fixed in 2.2.0.4. From the release notes:

Kubernetes: Persistent volumes created by claims are now stored in the virtual machine. Fixes docker/for-win#5665.
